Include a NavigationActivation object in the 'page swap' event.

mattwoodrow opened the pull request:
#### 392522f03118c2441c27bb304d97f4d70e342318
<pre>
Include a NavigationActivation object in the &apos;page swap&apos; event.
<a href="https://bugs.webkit.org/show_bug.cgi?id=278199">https://bugs.webkit.org/show_bug.cgi?id=278199</a>
&lt;<a href="https://rdar.apple.com/133993708">rdar://133993708</a>&gt;

Reviewed by Charlie Wolfe.

This creates a NavigationActivation object for the pageswap event (for same
origin navigations), as required by the spec.

The code for dispatching pageswap gets moved slightly earlier in the commit
process, to be before BackForwardCache::singleton().addIfCacheable. This code
mutates the global history, but also can fire the unload event if the old page
got put into the BF cache. Moving the pageswap event earlier guarantees that it
always happens before unload. It also makes the global history easier to reason
about, even if it doesn&apos;t match what the spec wants (as described in the inline
FIXME).

This code should work even with the Navigation API setting disabled, as its an
important feature for cross-document view-transitions (and demos are relying on
being able to access the new/old URLS).

* LayoutTests/imported/w3c/web-platform-tests/css/css-view-transitions/navigation/pageswap-push-navigation-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-view-transitions/navigation/pageswap-push-with-redirect-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-view-transitions/navigation/pageswap-replace-navigation-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-view-transitions/navigation/pageswap-traverse-navigation-no-bfcache.https-expected.txt:
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::dispatchPageswapEvent):
* Source/WebCore/dom/Document.h:
* Source/WebCore/loader/FrameLoader.cpp:
(WebCore::FrameLoader::commitProvisionalLoad):
(WebCore::FrameLoader::transitionToCommitted):
* Source/WebCore/loader/HistoryController.cpp:
(WebCore::HistoryController::restoreDocumentState):
(WebCore::HistoryController::goToItem):
(WebCore::HistoryController::updateForReload):
(WebCore::HistoryController::initializeItem):
(WebCore::HistoryController::createItem):
(WebCore::HistoryController::createNewItem):
(WebCore::HistoryController::updateCurrentItem):
* Source/WebCore/loader/HistoryController.h:
* Source/WebCore/page/Navigation.cpp:
(WebCore::Navigation::createForPageswapEvent):
* Source/WebCore/page/Navigation.h:

Canonical link: <a href="https://commits.webkit.org/282953@main">https://commits.webkit.org/282953@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/2aa111e01af4e53c830445a76e43b34ef27e981e

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/64748 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/44113 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/17344 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/68773 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/15355 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/66865 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/51894 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/15633 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/52053 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/10590 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/67814 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/40808 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/56021 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/32674 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/37473 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/13397 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/14230 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/59407 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/13726 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/70478 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/8693 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/13220 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/59381 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/8727 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/56106 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/59572 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/14287 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/7174 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/848 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/39925 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/41005 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/42185 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/40746 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

TingPing commented on Source/WebCore/page/Navigation.cpp:
```diff
@@ -147,6 +147,49 @@ void Navigation::updateForActivation(HistoryItem* previousItem, std::optional<Na
+        return NavigationHistoryEntry::create(scriptExecutionContext(), *item);
+    };
+
+    RefPtr<NavigationHistoryEntry> oldEntry = getOrCreateNavigationHistoryForHistory(frame() ? frame()->checkedHistory()->currentItem() : nullptr);
```
Does `currentEntry()` not return what you want?

mattwoodrow replied:
With the change you requested below, yes, it should.

TingPing commented on Source/WebCore/page/Navigation.cpp:
```diff
@@ -147,6 +147,49 @@ void Navigation::updateForActivation(HistoryItem* previousItem, std::optional<Na
+        return nullptr;
+
+    // The lookups into m_entries should succeed if the Navigation API is enabled, but
+    // we fallback to a new NavigationHistoryEntry if not so that the pageswap event can
```
Maybe make this more explicit, check if NavigationAPI is enabled and do the "right thing", then fallback otherwise.

mattwoodrow replied:
Done!

TingPing commented on Source/WebCore/page/Navigation.cpp:
```diff
@@ -147,6 +147,49 @@ void Navigation::updateForActivation(HistoryItem* previousItem, std::optional<Na
+        return nullptr;
+
+    bool isSameOrigin = SecurityOrigin::create(documentLoader->documentURL())->isSameOriginAs(window()->protectedDocument()->securityOrigin());
+    if (!isSameOrigin || !documentLoader->request().isSameSite())
```
I would add a comment here that you are checking `.isSameSite()` specifically to verify no cross-origin redirect has happened.

rwlbuis commented on Source/WebCore/loader/FrameLoader.cpp:
```diff
@@ -2246,6 +2247,36 @@ void FrameLoader::commitProvisionalLoad()
+                activation = domWindow->protectedNavigation()->createForPageswapEvent(newItem.get(), pdl.get());
+            }
+        }
+        document->dispatchPageswapEvent(canTriggerCrossDocumentViewTransition, activation.get());
```
Shouldn't the activation be moved?

rwlbuis commented on Source/WebCore/loader/FrameLoader.cpp:
```diff
@@ -2246,6 +2247,36 @@ void FrameLoader::commitProvisionalLoad()
+    if (RefPtr document = m_frame->document()) {
+        bool canTriggerCrossDocumentViewTransition = false;
+        RefPtr<NavigationActivation> activation;
+        if (pdl) {
```
Not your fault, but IMHO the abbreviation is not nice.

mattwoodrow replied:
Agreed. It's used in a lot of places in this function though, so I won't change it as part of this PR.

nt1m commented on Source/WebCore/dom/Document.h:
```diff
@@ -1383,7 +1384,7 @@ class Document
     void dispatchPageshowEvent(PageshowEventPersistence);
     void dispatchPagehideEvent(PageshowEventPersistence);
-    void dispatchPageswapEvent(bool canTriggerCrossDocumentViewTransition);
+        void dispatchPageswapEvent(bool canTriggerCrossDocumentViewTransition, RefPtr<NavigationActivation>&&);
```
nit: unintentional indentation

nt1m commented on Source/WebCore/page/Navigation.cpp:
```diff
@@ -147,6 +147,44 @@ void Navigation::updateForActivation(HistoryItem* previousItem, std::optional<Na
+    RefPtr<NavigationHistoryEntry> oldEntry;
+    if (frame()->document() && frame()->document()->settings().navigationAPIEnabled())
+        oldEntry = currentEntry();
+    else if (HistoryItem* currentItem = frame()->checkedHistory()->currentItem())
```
```suggestion
    else if (auto* currentItem = frame()->checkedHistory()->currentItem())
```

achristensen07 replied:
RefPtr currentItem?

charliewolfe commented on Source/WebCore/loader/FrameLoader.cpp:
```diff
@@ -2246,6 +2247,36 @@ void FrameLoader::commitProvisionalLoad()
+                // FIXME: The NavigationActivation for pageswap should be created after the global
+                // history update, but before the unload event (which might be delayed). Those steps
+                // are currently intertwined, so this creates a fake/detached new history entry to
+                // use for this purpose.
```
Is it difficult to separate the unload event from `addIfCacheable()`? Maybe this is okay for now, but it seems very weird to have this situation where we need to create a detached history item with the provisional loader.

mattwoodrow replied:
The answer is that I really don't know at all :) 

Some loads result in the unload event happening deep inside `addIfCacheable`, others happen when we call `transitionToCommitted` into `closeURL()` (which is what the comments say is the expected time).

This code has all sorts of interesting early exit paths where calling a JS event results in the navigation state being changed recursively, and it's really hard to reason about all the possible fallout, especially if there are nested <iframes>.

I couldn't agree more that it's weird, and would be good to unwind this code to match the spec description better, but that seems like a large and danger-fraught task :)

charliewolfe commented on Source/WebCore/loader/HistoryController.h:
```diff
@@ -94,12 +94,14 @@ class HistoryController final : public CanMakeCheckedPtr<HistoryController> {
 
     void setDefersLoading(bool);
 
+    Ref<HistoryItem> createNewItem(HistoryItemClient&, RefPtr<DocumentLoader>);
```
DocumentLoader parameter should just be a pointer

charliewolfe replied:
Also maybe rename this function to overload `HistoryController::createItem`? Or name `createItemWithLoader` or something.

charliewolfe commented on Source/WebCore/loader/FrameLoader.cpp:
```diff
@@ -2246,6 +2247,36 @@ void FrameLoader::commitProvisionalLoad()
+                RefPtr<HistoryItem> newItem;
+                if (*navigationAPIType != NavigationNavigationType::Reload) {
+                    RefPtr page = frame->page();
+                    newItem = frame->checkedHistory()->createNewItem(page->historyItemClient(), pdl);
```
page may need a null check

charliewolfe commented on Source/WebCore/loader/HistoryController.cpp:
```diff
@@ -741,11 +740,24 @@ void HistoryController::initializeItem(HistoryItem& item)
-    
+
+    RefPtr frame = dynamicDowncast<LocalFrame>(m_frame.ptr());
+    if (frame) {
```
nit - could restrict scope here
```suggestion
    if (RefPtr frame = dynamicDowncast<LocalFrame>(m_frame.ptr())) {
```

mattwoodrow committed:
Include a NavigationActivation object in the 'page swap' event.
https://bugs.webkit.org/show_bug.cgi?id=278199
<rdar://133993708>

Reviewed by Charlie Wolfe.

This creates a NavigationActivation object for the pageswap event (for same
origin navigations), as required by the spec.

The code for dispatching pageswap gets moved slightly earlier in the commit
process, to be before BackForwardCache::singleton().addIfCacheable. This code
mutates the global history, but also can fire the unload event if the old page
got put into the BF cache. Moving the pageswap event earlier guarantees that it
always happens before unload. It also makes the global history easier to reason
about, even if it doesn't match what the spec wants (as described in the inline
FIXME).

This code should work even with the Navigation API setting disabled, as its an
important feature for cross-document view-transitions (and demos are relying on
being able to access the new/old URLS).

* LayoutTests/imported/w3c/web-platform-tests/css/css-view-transitions/navigation/pageswap-push-navigation-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-view-transitions/navigation/pageswap-push-with-redirect-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-view-transitions/navigation/pageswap-replace-navigation-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-view-transitions/navigation/pageswap-traverse-navigation-no-bfcache.https-expected.txt:
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::dispatchPageswapEvent):
* Source/WebCore/dom/Document.h:
* Source/WebCore/loader/FrameLoader.cpp:
(WebCore::FrameLoader::commitProvisionalLoad):
(WebCore::FrameLoader::transitionToCommitted):
* Source/WebCore/loader/HistoryController.cpp:
(WebCore::HistoryController::restoreDocumentState):
(WebCore::HistoryController::goToItem):
(WebCore::HistoryController::updateForReload):
(WebCore::HistoryController::initializeItem):
(WebCore::HistoryController::createItem):
(WebCore::HistoryController::createNewItem):
(WebCore::HistoryController::updateCurrentItem):
* Source/WebCore/loader/HistoryController.h:
* Source/WebCore/page/Navigation.cpp:
(WebCore::Navigation::createForPageswapEvent):
* Source/WebCore/page/Navigation.h:

Canonical link: https://commits.webkit.org/282953@main
